lib/bolt/pal.rb

```diff
@@ -145,8 +145,22 @@ def in_bolt_compiler
             end
           rescue Bolt::Error => e
             e
+          rescue Puppet::DataBinding::LookupError => e
+            if e.issue_code == :HIERA_UNDEFINED_VARIABLE
+              message = "Interpolations are not supported in lookups outside of an apply block: #{e.message}"
+              PALError.new(message)
+            else
+              PALError.from_preformatted_error(e)
+            end
           rescue Puppet::PreformattedError => e
-            PALError.from_preformatted_error(e)
+            if e.issue_code == :UNKNOWN_VARIABLE &&
+               (match = /(?<var>facts|trusted|server_facts|settings)/.match(e.message))
```

Comment:
This currently matches variables that contain these variable names as substrings. Even more confusingly, the substring it matches is the only part it prints in the error message.

```
plan test() {
  return $factses
}
```

```
{
  "kind": "bolt/pal-error",
  "msg": "Evaluation Error: Variable 'facts' is not available in the current scope unless explicitly defined. (file: /Users/nicklewis/projects/bolt/Boltdir/site/test/plans/init.pp, line: 2, column: 10)",
  "details": {
  }
}
```

It looks like the variable name may be available as `e.arguments[:name]`? I'm trying to trace the code paths to figure out _how_ that becomes the case, but it's... complex. But if that's reliable, then we can avoid using a regex entirely.

Reply:
That's helpful. Updated to use the argument from the error for now.